File: 일자별/2026년/2026.04/2026.04.18/베스트앨범.py
genres,plays = ["classic", "pop", "classic", "classic", "pop"],[500, 600, 150, 800, 2500]

from collections import defaultdict
def solution(genres,plays) :
    answer = []
    genre_dic = defaultdict(int)
    music_dic = defaultdict(list)
    for idx,genre in enumerate(genres) :
        music_dic[genre].append((plays[idx],-idx))
        genre_dic[genre] += plays[idx]
    
    sorted_genres = sorted(genre_dic.items(), key = lambda x : x[1] ,reverse = True)

    for genre,_ in sorted_genres : 
        music_dic[genre].sort(reverse = True)
        cnt = 0
        for music in music_dic[genre] :
            answer.append(-music[1])
            cnt += 1
            if cnt == 2 :
                break

    return answer

print(solution(genres,plays))

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
  const answer = [];
  const playCnt = plays.length;
  const genreTotal = new Map();
  const genrePerPlays = new Map();

  for (let i = 0; i < playCnt; i++) {
    genreTotal.set(genres[i], (genreTotal.get(genres[i]) || 0) + plays[i]);
    if (!genrePerPlays.has(genres[i])) genrePerPlays.set(genres[i], []);
    genrePerPlays.get(genres[i]).push([plays[i], i]);
  }
  sortedGenre = [...genreTotal].sort((a, b) => b[1] - a[1]);

  for (const list of genrePerPlays.values()) {
    list.sort((a, b) => b[0] - a[0]);
  }

  for (const sg of sortedGenre) {
    for (const gpp of genrePerPlays.get(sg[0]).slice(0, 2)) {
      answer.push(gpp[1]);
    }
  }
  return answer;
}